javagen/src/main/java/com/azure/autorest/mapper/ModelMapper.java

```diff
@@ -206,14 +209,48 @@ public ClientModel map(ObjectSchema compositeType) {
             if (isPolymorphic) {
                 String discriminatorSerializedName = SchemaUtil.getDiscriminatorSerializedName(compositeType);
                 // Only escape the discriminator if the model will be flattened.
-                builder.polymorphicDiscriminator(needsFlatten
-                        ? discriminatorSerializedName.replace(".", "\\\\.")
-                        : discriminatorSerializedName);
+                String polymorphicDiscriminator = needsFlatten
+                    ? discriminatorSerializedName.replace(".", "\\\\.")
+                    : discriminatorSerializedName;
+
+                builder.polymorphicDiscriminator(polymorphicDiscriminator);
+
+                // Given the discriminator value is being passed if this is the terminal type
+                // add a read-only property to the class which is the discriminator type.
+                if (settings.isDiscriminatorPassedToChildDeserialization() && !hasChildren) {
```

Comment:
Extract this out into a separate method to allow fluent plugin to override this behavior if necessary.